python/Ganga/GPIDev/Base/Proxy.py

```diff
@@ -561,16 +570,25 @@ def GPIProxyObjectFactory(_obj):
         from Ganga.Core.exceptions import GangaException
         raise GangaException("%s is NOT a Proxyable object" % type(_obj))
 
-    if hasattr(_obj, proxyObject):
-        return getattr(_obj, proxyObject)
+    ## This is defined within Objects.py, we could probably store this elsehere
+    ## (We probably do) but as this is guaranteed to be accessible for GangaObjects I will use this
+    this_class = getattr(stripProxy(_obj).__class__, proxyClass)
 
-    if not hasattr(_obj.__class__, proxyClass):
-        from Ganga.Core.exceptions import GangaException
-        raise GangaException("%s is NOT a Proxyable object" % type(_obj))
+    ## Would LOVE to avoid having to perform all of these checks to get or wrap a class in a proxy here but
+    ## multiple different BAD states are currently passed to this function
+    ## In future I would like to see this fixed elsewhere and reduce the logic in this section of code
+    ## As of Ganga 6.1.15 this is needed to put objects into a semi-coherent state
 
-    proxy_class = getattr(_obj.__class__, proxyClass)
+    if isType(_obj, this_class):
+        if isProxy(_obj):
+            return _obj
+        elif not isProxy(_obj):
+            pass
+    else:
+        pass
 
-    return proxy_class(_existing=_obj)
+    proxy_class = this_class(_proxy_impl_obj_to_wrap=_obj)
+    return proxy_class
```

Comment:
I've changed the function to always add a proxy to all GangaObject derivatives whilst returning Proxies which are already assigned to objects. As for where the proxyClass constructor is pulled from we're both using GangaObject ```__class__``` structure to get the corresponding proxy_class (this_class) which is what I'm hoping to avoid.